File: frontend/src/api/jobs.ts
/**
 * Typed client for the Jobs framework (Hardening Phase H1).
 *
 * The notification bell polls ``listActiveJobs`` every 3-5s while
 * any job is in-flight (or the bell dropdown is open). The shapes
 * here mirror ``backend/app/services/jobs_service.serialize_job``
 * exactly — keep in lockstep.
 */

import api from './client';


export type JobStatus =
    | 'queued'
    | 'running'
    | 'succeeded'
    | 'failed'
    | 'cancelled';


export interface Job {
    id: number;
    kind: string;
    title: string;
    status: JobStatus;
    progress: number | null;
    progress_message: string | null;
    project_id: number | null;
    user_id: number | null;
    params: Record<string, unknown>;
    result: Record<string, unknown> | null;
    error: string | null;
    queued_at: string | null;
    started_at: string | null;
    completed_at: string | null;
    dismissed_at: string | null;
}


export interface ActiveJobsResponse {
    count: number;
    jobs: Job[];
}


export async function listActiveJobs(opts: {
    projectId?: number | null;
    includeRecentlyCompleted?: boolean;
    limit?: number;
} = {}): Promise<ActiveJobsResponse> {
    const params: Record<string, string> = {};
    if (opts.projectId !== undefined && opts.projectId !== null) {
        params.project_id = String(opts.projectId);
    }
    if (opts.includeRecentlyCompleted !== undefined) {
        params.include_recently_completed = String(opts.includeRecentlyCompleted);
    }
    if (opts.limit !== undefined) {
        params.limit = String(opts.limit);
    }
    const query = new URLSearchParams(params).toString();
    const url = query ? `/jobs/active?${query}` : '/jobs/active';
    const resp = await api.get<ActiveJobsResponse>(url);
    return resp.data;
}


export async function getJob(jobId: number): Promise<Job> {
    const resp = await api.get<Job>(`/jobs/${jobId}`);
    return resp.data;
}


export async function dismissJob(jobId: number): Promise<Job> {
    const resp = await api.post<Job>(`/jobs/${jobId}/dismiss`);
    return resp.data;
}


export async function cancelJob(jobId: number): Promise<Job> {
    const resp = await api.post<Job>(`/jobs/${jobId}/cancel`);
    return resp.data;
}

File: backend/app/api/jobs.py
"""Jobs API routes (Hardening Phase H1).

Read-only surface that the frontend's notification bell polls.
Mutations (start, cancel, dismiss) live here too so the UI doesn't
need to talk to two routers.
"""

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.services.jobs_service import (
    cancel_job,
    dismiss_job,
    get_job,
    list_active_jobs,
    serialize_job,
    serialize_job_with_live_metrics,
)


router = APIRouter(prefix="/jobs", tags=["Jobs"])


@router.get("/active")
async def get_active_jobs(
    project_id: int | None = None,
    include_recently_completed: bool = True,
    limit: int = 50,
    db: AsyncSession = Depends(get_db),
):
    """List in-flight + recently-completed jobs.

    Frontend's notification bell polls this every 3-5 seconds while
    the user has any tab open. Optionally scope by ``project_id``
    when the bell is shown on a project page; pass ``project_id=null``
    (omit) for the global "all my jobs" view.
    """
    jobs = await list_active_jobs(
        db,
        project_id=project_id,
        include_recently_completed=include_recently_completed,
        limit=max(1, min(limit, 200)),
    )
    return {
        "count": len(jobs),
        # Bell consumers get the live-metrics enrichment so the
        # sparkline can render during a training run.
        "jobs": [serialize_job_with_live_metrics(j) for j in jobs],
    }


@router.get("/{job_id}")
async def get_job_by_id(
    job_id: int,
    db: AsyncSession = Depends(get_db),
):
    """Fetch a single job by id. Useful for the "Open" deep-link
    from a notification when the bell payload is stale."""
    job = await get_job(db, job_id)
    if job is None:
        raise HTTPException(404, f"Job {job_id} not found")
    return serialize_job_with_live_metrics(job)


@router.post("/{job_id}/dismiss")
async def dismiss(
    job_id: int,
    db: AsyncSession = Depends(get_db),
):
    """Hide a completed job from the notification bell.
    In-flight jobs can't be dismissed — caller should cancel."""
    try:
        job = await dismiss_job(db, job_id)
    except ValueError as exc:
        raise HTTPException(400, str(exc)) from exc
    if job is None:
        raise HTTPException(404, f"Job {job_id} not found")
    return serialize_job(job)


@router.post("/{job_id}/cancel")
async def cancel(
    job_id: int,
    db: AsyncSession = Depends(get_db),
):
    """Request job cancellation. Cooperative — runner has to honor
    it at its next checkpoint. For runners that don't yet
    cooperate, this is a "stop showing me" flag (the work
    still completes server-side)."""
    job = await cancel_job(db, job_id)
    if job is None:
        raise HTTPException(404, f"Job {job_id} not found")
    return serialize_job(job)
